### ISSUE TITLE
Need code for selecting top-m sentences (principal sentence selection)

### ISSUE THREAD

**Issue by agenius5** [google-research/pegasus#122]
Hii, as mentioned in the paper ROUGE1- F measure is used to select sentences for masking. I can't find the code to do the same. Please point to the file in which this functionality is implemented. 

I implemented similar functionality with nltk tokenizer and rouge_score library and I am getting precision, recall and f-measure for each sentence. So, am I doing it right? 

Moreover, as we know the max input length is 1024 tokens so can we use this sentence ranking to get our input within 1024 token limit by removing less important sentences?

**Comment by JingqingZ** [comment]
> Hii, as mentioned in the paper ROUGE1- F measure is used to select sentences for masking. I can't find the code to do the same. Please point to the file in which this functionality is implemented.

https://github.com/google-research/pegasus/blob/939830367bcf411193d2b5eca2f2f90f3f9260ca/pegasus/ops/pretrain_parsing_ops.cc#L294

> Moreover, as we know the max input length is 1024 tokens so can we use this sentence ranking to get our input within 1024 token limit by removing less important sentences?

Technically, yes, you can use the ranking to select more informative sentences.

**Comment by agenius5** [comment]
@JingqingZ Hii,

I'll come straight to the point. What I am trying to do is extract important sentences from the long inputs before feeding it to the model for summarization. I believe this should work better than truncation.

As an extension I am also thinking of training the model the same way. During fine-tuning long sequences are truncated so I was thinking of putting a "extractive approach" which selects important senteces (with in encoder's input limit) for fine-tuning. After fine-tuning this way I want to evaluate to see how it performs.

What do you think of this apporach?

**Comment by JingqingZ** [comment]
I remember there are several papers talking about similar process. Basically, extract most informative sentences from a long document and then generate abstractive summaries. May be worth exploring the related papers.

**Comment by agenius5** [comment]
@JingqingZ 

I know this process has been widely used. I am trying to apply the same things on PEGASUS during fine-tuninig phase as a part of my project on Abstractive text summarization. To do that I need to modify the dataset and fine-tune on this modified dataset.
Other optioned that I tried was to find where exactly each example is fed to the model for fine-tuning and do some pre-processing right there.

I studied the code and found these two methods in public_parsing_ops.py:

`def encode(text: tf.Tensor, max_len: int, vocab_filename: str, encoder_type: str):`

` def parser(input_dic):`

In parser function, from `input_dic["inputs"]` and in parser function, from `text' I tried to get the input sequence so I can apply a extractive method on it right there but I couldn't get the input sequnce. I tried to access the numpy() value from tensor but it didn't work due to some reasons.

So can you please tell me if it possible to modify the input sequence that is fed to the model for fine-tuning. If yes, the where I do that in source code.

I'm stuck badly. It'll be a great help.

Thanks.

**Comment by JingqingZ** [comment]
> So can you please tell me if it possible to modify the input sequence that is fed to the model for fine-tuning. If yes, the where I do that in source code. If No, then do I have to built my own dataset?

https://github.com/google-research/pegasus/blob/13fcf2b1191e0df950436c82c33d672c1447f5ff/pegasus/data/parsers.py#L48

You may modify the logic here to process inputs.

>  I tried to get the input sequence so I can apply a extractive method on it right there but I couldn't get the input sequnce.

Can you please elaborate? If you're using one of the existing supervised datasets, the parser should be able to get the inputs and targets.

**Comment by agenius5** [comment]
@JingqingZ 

> I tried to get the input sequence so I can apply a extractive method on it right there but I couldn't get the input sequnce.

**Comment by agenius5** [comment]
> I tried to get the input sequence so I can apply a extractive method on it right there but I couldn't get the input sequnce.

Actually I wanted to get the input sequence in `str` format in order to apply the _extractive part_. In 'def parser(input_dic)` function it was not possible to get the input sequnce( in `str` format) from the input sequcne tensor (which we get with 'input_dic["inputs"]`). However, I was able to apply to do it with 'tf.py_function' in 'def parser(input_dic)` function.

Basically, I want to fine-tune the model _extract important sentences then summarize abstractively_ way.

I studied the code and realized that we can also pas .tfrecord files as datasets. So, what I did is load the public tfds -> take _x_ number of examples -> get the input sequence to apply _extractive part_ and keep the summary same -> create a new dataset with tfrecord format ->fine-tune the model. 

And I was able to fine-tune this way on AESLC dataset. Now, i'm evaluating this model. I have a few more questions though.
